### PR TITLE
Killer move

### PR DESCRIPTION
**Pull request by yibeili**

### CODE REVIEW

File: sporkfish/searcher/move_ordering.py [hunk 1]
```diff
@@ -69,3 +70,38 @@ def evaluate(self, board: Board, move: chess.Move) -> float:
             ]
         else:
             return 0
+
+
+class KillerMoveHeuristic(MoveOrder):
+    def __init__(self, max_depth: int) -> None:
+        # Initialize a list to store up to two killer moves for each depth
+        # Using a fixed size per depth for simplicity; could be dynamic based on actual usage
+        self.killer_moves: List[List[Optional[chess.Move]]] = [
```
Comment (KYLChiu) [inline sporkfish/searcher/move_ordering.py:79]: Could we make this protected please?

File: sporkfish/searcher/move_ordering.py [hunk 2]
```diff
@@ -69,3 +70,38 @@ def evaluate(self, board: Board, move: chess.Move) -> float:
...
+        """
+        killers = self.killer_moves[depth]
+        if move not in killers:
+            # Shift the existing killer moves down and add the new one at the front
+            killers.pop()
+            killers.insert(0, move)
+
+    def evaluate(self, board: Board, move: chess.Move, depth: int) -> float:
+        killers = self.killer_moves[depth]
+        if move in killers:
```
Comment (KYLChiu) [inline sporkfish/searcher/move_ordering.py:100]: On second thought 

return 1 if move in killers else 0

looks cleaner to me. Sorry!

File: LICENSE [hunk 3]
```diff
@@ -19,3 +19,6 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
+
```
Comment (KYLChiu) [inline LICENSE:22]: Do you mind getting rid of these please?

**Review by KYLChiu — COMMENTED** [review]
Needs to fix the tests and add some as discussed, but otherwise LGTM!

File: sporkfish/searcher/move_ordering.py [hunk 4]
```diff
@@ -69,3 +70,38 @@ def evaluate(self, board: Board, move: chess.Move) -> float:
...
+            move (chess.Move): The move to be added as a killer move.
+            depth (int): The depth at which the move proved to be a killer move.
+        """
+        killers = self.killer_moves[depth]
+        if move not in killers:
+            # Shift the existing killer moves down and add the new one at the front
+            killers.pop()
+            killers.insert(0, move)
+
+    def evaluate(self, board: Board, move: chess.Move, depth: int) -> float:
```
Comment (KYLChiu) [inline sporkfish/searcher/move_ordering.py:98]: Please add a docstring here too

File: sporkfish/searcher/move_ordering.py [hunk 5]
```diff
@@ -69,3 +70,38 @@ def evaluate(self, board: Board, move: chess.Move) -> float:
...
+        ]
+
+    def add_killer_move(self, move: chess.Move, depth: int) -> None:
+        """
+        Add a move to the killer moves list for a given depth, ensuring it's not already present.
+        If the move is new, it replaces the oldest killer move.
+
+        Parameters:
+            move (chess.Move): The move to be added as a killer move.
+            depth (int): The depth at which the move proved to be a killer move.
```
Comment (KYLChiu) [inline sporkfish/searcher/move_ordering.py:90]: The doc strong is correct but we need *Sphinx* docs. Could you change these please?

File: sporkfish/searcher/move_ordering.py (hunk 1 above)
Comment (ccjeremylo) [reply sporkfish/searcher/move_ordering.py:79]: Or even make it private? Actually wonder why we never use private members in `Sporkfish` - suspect it was just us being lazy and didn't bother to understand difference between private and protected?

Sometimes, we do want protected members, especially when we need those members in child classes. But say in this case, I believe private accessibility is what we what.

File: sporkfish/searcher/searcher_factory.py [hunk 6]
```diff
@@ -13,7 +13,7 @@ class SearcherFactory:
     """
 
     @staticmethod
-    def _build_moveorder(order_type: MoveOrderMode) -> MoveOrder:
+    def _build_moveorder(searcher_cfg: SearcherConfig) -> MoveOrder:
```
Comment (ccjeremylo) [inline sporkfish/searcher/searcher_factory.py:16]: What is the reason for this change? e.g. is it more user friendly, or is it some subtle performance thing?

[edit] ah ok, I see you need extra config info as part of this change...sorry was blind

File: sporkfish/searcher/searcher_factory.py [hunk 7]
```diff
@@ -23,8 +23,12 @@ def _build_moveorder(order_type: MoveOrderMode) -> MoveOrder:
         :rtype: MoveOrder
         :raises TypeError: If the specified order type is not supported.
         """
+        order_type = searcher_cfg.move_order_mode
+        max_depth = searcher_cfg.max_depth
```
Comment (ccjeremylo) [inline sporkfish/searcher/searcher_factory.py:27]: is `max_depth` only needed for `KILLER_MOVE`? if so suggest move it down to inside the `elif` statement

File: sporkfish/searcher/searcher_factory.py [hunk 8]
```diff
@@ -13,7 +13,7 @@ class SearcherFactory:
     """
 
     @staticmethod
-    def _build_moveorder(order_type: MoveOrderMode) -> MoveOrder:
+    def _build_moveorder(searcher_cfg: SearcherConfig) -> MoveOrder:
         """
         Build and return an instance of MoveOrder based on the specified order type.
```
Comment (ccjeremylo) [inline sporkfish/searcher/searcher_factory.py:18]: Think doc string needs updating

File: sporkfish/searcher/move_ordering.py (hunk 2 above)
Comment (ccjeremylo) [reply sporkfish/searcher/move_ordering.py:100]: Checking my understanding, so `killer_moves` evaluation simply checks if a move is a [killer](https://rustic-chess.org/search/ordering/killers.html), and gives that move a higher score?

So in fact, most non-killer-moves have very similar scores? How are those being ordered?

**Review by ccjeremylo — COMMENTED** [review]
Looking goodddd, especially given this is a first PR 😸 

Added some comments + sanity checking questions, nothing major. Also noticed CI is red, I think due to broken function calls in tests after changing the function signature - hopefully easy-ish to fix (?) 🤔 

Can see @KYLChiu already suggested some tests to add as well 😛

**Comment by ccjeremylo** [comment]
Some general comments - usually it is nice to put some description of:

- what has been changed/added
- why there is such a change

This is for documentation purposes and also helps reviewer to get some context - of course there might be some duplication with the original issue, e.g. #4.

**Comment by ccjeremylo** [comment]
Also, I guess we need to add configs for this change? I don't think this has been done?

Are we going to default to `MVV_LVA` or `KILLER_MOVE`? I suspect this discussion is related to #114

File: sporkfish/searcher/searcher_factory.py (hunk 7 above)
Comment (KYLChiu) [reply sporkfish/searcher/searcher_factory.py:27]: I missed, ty :)

File: sporkfish/searcher/move_ordering.py (hunk 1 above)
Comment (KYLChiu) [reply sporkfish/searcher/move_ordering.py:79]: https://google.github.io/styleguide/pyguide.html

Section Apparently Google likes it with the guidance of the Python creator

Comment (KYLChiu) [reply sporkfish/searcher/move_ordering.py:79]: https://google.github.io/styleguide/pyguide.html

Section Apparently Google likes it with the guidance of the Python creator

Comment (KYLChiu) [reply sporkfish/searcher/move_ordering.py:79]: Section 3.16, https://google.github.io/styleguide/pyguide.html

Apparently Google uses it with guidance from Python creator, I'm in no position to argue :) Though I recall there might be be a non stylistic reason with private members and name mangling, but can't find the reference... do let me know if you find out

File: sporkfish/searcher/move_ordering.py [hunk 9]
```diff
@@ -69,3 +70,38 @@ def evaluate(self, board: Board, move: chess.Move) -> float:
...
+        """
+        killers = self.killer_moves[depth]
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/searcher/move_ordering.py:92]: Good luck Yibei!

**Comment by KYLChiu** [comment]
> Also, I guess we need to add configs for this change? I don't think this has been done?
> 
> 
> 
> Are we going to default to `MVV_LVA` or `KILLER_MOVE`? I suspect this discussion is related to #114 

No, let's keep it as is for now. The naive idea is that usually captures (MVV_LVA) are more valuable than quiet (killer) moves. Later on we'll have a composite move ordering class/methodology which combines these two and any future move ordering classes.

File: sporkfish/searcher/move_ordering.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/searcher/move_ordering.py:92]: Interesting, TIL.

Unfortunately this is an inherently mutable thing, it's similar to the transposition table/cache for killer moves. 

No objections to switching to directly updating it...

File: sporkfish/searcher/move_ordering.py [hunk 10]
```diff
@@ -10,11 +10,12 @@
...
+    KILLER_MOVE = "KILLER_MOVE"
 
 
 class MoveOrder(ABC):
     @abstractmethod
-    def evaluate(self, board: Board, move: chess.Move) -> float:
+    def evaluate(self, board: Board, move: chess.Move, depth: int) -> float:
         """
         Abstract method to evaluate the desirability of a move in a given board position.
         Higher values indicate more desirable moves.
```
[the remark this answers is not in the document]
Comment (yibeili) [reply sporkfish/searcher/move_ordering.py:21]: Is this description okay?
```
:param depth: The current depth at which the move ordering is considered.
:type depth: int
```

File: sporkfish/searcher/minimax.py [hunk 11]
```diff
@@ -65,7 +65,7 @@ def __init__(
     def evaluator(self) -> Evaluator:
         return self._evaluator
 
-    def _ordered_moves(self, board: Board, legal_moves: Any) -> Any:
+    def _ordered_moves(self, board: Board, legal_moves: Any, depth: int) -> Any:
         """
         Order the given legal moves from best to worst based on a move ordering heuristic.
```
[the remark this answers is not in the document]
Comment (yibeili) [reply sporkfish/searcher/minimax.py:71]: Is this description okay?
```
:param depth: The current depth at which the move ordering is considered.
:type depth: int
```

[the remark this answers is not in the document]
Comment (ccjeremylo) [reply sporkfish/searcher/minimax.py:71]: Yep LGTM (looks good to me)

File: LICENSE (hunk 3 above)
Comment (KYLChiu) [reply LICENSE:22]: Looks like they're still here!

File: sporkfish/searcher/move_ordering.py (hunk 1 above)
Comment (yibeili) [reply sporkfish/searcher/move_ordering.py:79]: Done!

File: sporkfish/searcher/move_ordering.py (hunk 2 above)
Comment (yibeili) [reply sporkfish/searcher/move_ordering.py:100]: Killer move heuristic is not great move ordering by itself. It is better when combined with mvvlva. We will redesign the structure when we combine them later. #114

Comment (yibeili) [reply sporkfish/searcher/move_ordering.py:100]: worked fine with mypy
```
sporkfish % mypy ./sporkfish/searcher/move_ordering.py
Success: no issues found in 1 source file
```

File: sporkfish/searcher/move_ordering.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (yibeili) [reply sporkfish/searcher/move_ordering.py:92]: done!

File: sporkfish/searcher/move_ordering.py [hunk 12]
```diff
@@ -69,3 +73,50 @@ def evaluate(self, board: Board, move: chess.Move) -> float:
...
         else:
             return 0
+
+
+class KillerMoveHeuristic(MoveOrder):
+    def __init__(self, max_depth: int) -> None:
+        # Store up to two killer moves for each depth
+        # Using a fixed size per depth for simplicity; could be dynamic based on actual usage
+        self._killer_moves: List[List[Optional[chess.Move]]] = [
+            [None, None] for _ in range(max_depth + 1)
```
Comment (KYLChiu) [inline sporkfish/searcher/move_ordering.py:83]: NB: fixed max_depth + 1, because it should include max_depth...

File: sporkfish/searcher/move_ordering.py [hunk 13]
```diff
@@ -46,17 +48,19 @@ def __init__(self) -> None:
...
-            board (Board): The chess board.
-            move (chess.Move): The capturing move.
-
-        Returns:
-            int: The heuristic value of the capturing move based on the value of the captured piece.
+        :param board: The current state of the chess board.
+        :type board: Board
+        :param move: The move to be evaluated.
+        :type move: chess.Move
+        :param depth: The current depth at which the move ordering is considered.
```
Comment (ccjeremylo) [inline sporkfish/searcher/move_ordering.py:60]: nice tyty

File: sporkfish/searcher/move_ordering.py [hunk 14]
```diff
@@ -23,7 +24,8 @@ def evaluate(self, board: Board, move: chess.Move) -> float:
         :type board: Board
         :param move: The move to be evaluated.
         :type move: chess.Move
-
+        :param depth: The current depth at which the move ordering is considered.
```
Comment (ccjeremylo) [inline sporkfish/searcher/move_ordering.py:27]: 👍

File: sporkfish/searcher/move_ordering.py (hunk 12 above)
Comment (ccjeremylo) [reply sporkfish/searcher/move_ordering.py:83]: makes sense - nice catch

File: tests/test_searcher.py [hunk 15]
```diff
@@ -442,10 +452,54 @@ def test_sorting_legal_moves(
...
             mo = MvvLvaHeuristic()
-            score = mo.evaluate(board, move)
+            score = mo.evaluate(board, move, 0)
             assert score == move_scores[num_moves - i - 1]
+
+
+@pytest.mark.parametrize(
+    ("fen_string", "move_scores"),
+    [
+        ("k7/3q4/8/K7/2N5/8/8/8 w - - 1 34", [0, 0, 0, 0, 1, 0, 0, 0, 0, 0]),
```
Comment (ccjeremylo) [inline tests/test_searcher.py:467]: Where did you get this fen string from? 

[Very minor] Suggest using one of the fen string we have in the helper module? Easier to read and interpret

File: tests/test_searcher.py [hunk 16]
```diff
@@ -442,10 +452,54 @@ def test_sorting_legal_moves(
...
+    ],
+)
+class TestKillerMoveHeuristic:
```
Comment (ccjeremylo) [inline tests/test_searcher.py:470]: Ty for adding this!

File: tests/test_searcher.py [hunk 17]
```diff
@@ -442,10 +452,54 @@ def test_sorting_legal_moves(
...
+        mo = KillerMoveHeuristic(1)
+        mo.add_killer_move(chess.Move.from_uci("c4b6"), 1)
+
+        for i, move in enumerate(all_moves):
+            score = mo.evaluate(board, move, 1)
+            assert score == move_scores[i]
+
+    def test_sorting_legal_moves(self, fen_string: str, move_scores: list[int]) -> None:
+        """
+        Test sorting of legal moves by mvv_lva_heuristic
```
Comment (ccjeremylo) [inline tests/test_searcher.py:492]: Doc string needs updating

**Review by ccjeremylo — COMMENTED** [review]
Very minor points - should be easy to fix!!

File: tests/test_searcher.py (hunk 15 above)
Comment (KYLChiu) [reply tests/test_searcher.py:467]: I got it off a position I made. I feel it's more suited to the killer move setting - forking king/queen

File: tests/test_searcher.py [hunk 18]
```diff
@@ -442,10 +452,54 @@ def test_sorting_legal_moves(
...
+    [
+        ("k7/3q4/8/K7/2N5/8/8/8 w - - 1 34", [0, 0, 0, 0, 1, 0, 0, 0, 0, 0]),
+    ],
+)
+class TestKillerMoveHeuristic:
+    def test_ordered_moves_end_game(
+        self, fen_string: str, move_scores: list[int]
+    ) -> None:
+        """
+        Test mvv lva heuistic on an end game board
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply tests/test_searcher.py:475]: I removed the docstring, I think it's fairly clear from then names of the classes/functions - they don't really add much value?

File: tests/test_searcher.py (hunk 17 above)
Comment (KYLChiu) [reply tests/test_searcher.py:492]: Removed

File: tests/test_searcher.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (ccjeremylo) [reply tests/test_searcher.py:475]: yep sounds good to me

File: tests/test_searcher.py (hunk 15 above)
Comment (ccjeremylo) [reply tests/test_searcher.py:461]: Suggest an inline comment??

File: sporkfish/searcher/searcher_factory.py [hunk 19]
```diff
@@ -13,18 +13,23 @@ class SearcherFactory:
...
         :return: An instance of MoveOrder.
         :rtype: MoveOrder
         :raises TypeError: If the specified order type is not supported.
         """
+        order_type = searcher_cfg.move_order_mode
+
         if order_type is MoveOrderMode.MVV_LVA:
             return MvvLvaHeuristic()
+        elif order_type is MoveOrderMode.KILLER_MOVE:
+            max_depth = searcher_cfg.max_depth
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/searcher/searcher_factory.py]: `Updated`

File: tests/test_searcher.py [hunk 20]
```diff
@@ -430,22 +437,57 @@ def test_ordered_moves_end_game(
...
+            SearcherConfig(move_order_mode=MoveOrderMode.KILLER_MOVE)
+        )
+
+        mo: KillerMoveHeuristic = s._move_order
+        mo.add_killer_move(chess.Move.from_uci("c4b6"), 1)
+        legal_moves = s._ordered_moves(board, board.legal_moves, 1)
+
+        for i, move in enumerate(legal_moves):
+            score = mo.evaluate(board, move, 1)
+            assert score == (1 if i == 0 else 0)
```
Comment (ccjeremylo) [inline tests/test_searcher.py:494]: [Minor] This test only works for a unique board - I think using @pytest.mark.parametrize might not be good idea, as it suggest that we can add more inputs for testing

Not blocker

File: tests/test_searcher.py [hunk 21]
```diff
@@ -430,22 +437,57 @@ def test_ordered_moves_end_game(
...
+    ],
+)
+class TestKillerMoveHeuristic:
+    def test_ordered_moves_end_game(
+        self, fen_string: str, move_scores: list[int]
+    ) -> None:
+        board = init_board(fen_string)
+
+        all_moves = board.legal_moves
+        print(board.legal_moves)
```
Comment (ccjeremylo) [inline tests/test_searcher.py:472]: [Minor] suggest removing this

File: sporkfish/searcher/negamax.py [hunk 22]
```diff
@@ -88,6 +87,8 @@ def _negamax(
             alpha = max(alpha, value)
```
[the remark this answers is not in the document]
Comment (KYLChiu) [reply sporkfish/searcher/negamax.py:88]: OK

File: tests/test_searcher.py [hunk 23]
```diff
@@ -430,22 +437,57 @@ def test_ordered_moves_end_game(
...
+        num_moves = len(move_scores)
+        assert len(list(all_moves)) == num_moves
+        mo = KillerMoveHeuristic(1)
+        mo.add_killer_move(chess.Move.from_uci("c4b6"), 1)
+
+        for i, move in enumerate(all_moves):
+            score = mo.evaluate(board, move, 1)
+            assert score == move_scores[i]
```
Comment (ccjeremylo) [inline tests/test_searcher.py:480]: Nice!

File: tests/test_searcher.py [hunk 24]
```diff
@@ -339,9 +345,13 @@ def test_quiescence_depth_1_alpha(
 
 
 @pytest.fixture
-def _init_searcher(max_depth: int = 4) -> Searcher:
+def _init_searcher(
```
Comment (KYLChiu) [inline tests/test_searcher.py:348]: Re this: I have no idea how to use this properly. I can't call the function so I can't pass in arguments. Why is this a fixture instead of a pure function? I might be doing something wrong.

**Review by ccjeremylo — APPROVED** [review]
Left some v minor stylistic suggestions - not blocker.

Looks v good to me - approved! 😸

File: tests/test_searcher.py (hunk 21 above)
Comment (KYLChiu) [reply tests/test_searcher.py:472]: Will do next PR thank you

File: tests/test_searcher.py (hunk 20 above)
Comment (KYLChiu) [reply tests/test_searcher.py:494]: fair play, sounds g
